pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java

```diff
@@ -98,8 +104,36 @@ public String toExplainString() {
   protected void addRowsToRightTable(List<Object[]> rows) {
     assert _rightTable != null : "Right table should not be null when adding rows";
     for (Object[] row : rows) {
-      _rightTable.addRow(_rightKeySelector.getKey(row), row);
+      Object key = _rightKeySelector.getKey(row);
+      // Skip rows with null join keys - they should not participate in equi-joins per SQL standard
+      if (isNullKey(key)) {
+        // For RIGHT and FULL JOIN, we need to preserve null key rows for the final output
+        if (_nullKeyRightRows != null) {
+          _nullKeyRightRows.add(row);
+        }
+        continue;
+      }
+      _rightTable.addRow(key, row);
+    }
+  }
+
+  /**
+   * Check if a join key contains null values. In SQL standard, null keys should not match in equi-joins.
+   **/
+  private boolean isNullKey(Object key) {
+    if (key == null) {
+      return true;
     }
+    // For composite keys (Object[]), check if any component is null
+    if (key instanceof Object[]) {
```

Comment:
Isn't the key going to be an instance of `org.apache.pinot.core.data.table.Key` for the composite key case (see `MultiColumnKeySelector`)?

Reply:
Let's also add a unit test for this case.